perf(parser): faster checking for invalid modifiers

overlookmotel opened the pull request:
`Modifiers` contains 2 different representations of the modifiers list:

* `ModifierFlags` bitflags
* `Vec<Modifier>`

When checking modifiers, use `ModifierFlags` (which is really cheap), rather than iterating through the `Vec<Modifier>` (more expensive).

Only if the first check fails, then loop through the `Vec` to get the spans of the invalid modifiers, and generate diagnostics.

Because invalid modifiers are rare, move the logic for looping through the `Vec` into a `#[cold]` `#[inline(never)]` function. This should make it possible to inline `verify_modifiers` into all the call sites.

Produces a small perf improvement on parser benchmarks (+0.1% - +0.2%).

overlookmotel commented:
* **#15717** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15717?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/15717?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=15717).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Nov 16, 3:03 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Nov 16, 3:03 AM UTC**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
perf(parser): faster checking for invalid modifiers (#15717)

`Modifiers` contains 2 different representations of the modifiers list:

* `ModifierFlags` bitflags
* `Vec<Modifier>`

When checking modifiers, use `ModifierFlags` (which is really cheap), rather than iterating through the `Vec<Modifier>` (more expensive).

Only if the first check fails, then loop through the `Vec` to get the spans of the invalid modifiers, and generate diagnostics.

Because invalid modifiers are rare, move the logic for looping through the `Vec` into a `#[cold]` `#[inline(never)]` function. This should make it possible to inline `verify_modifiers` into all the call sites.

Produces a small perf improvement on parser benchmarks (+0.1% - +0.2%).
